src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenRefReadonlyParameterTests.cs

```diff
@@ -0,0 +1,593 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics;
+
+// PROTOTYPE: Move to RefReadonlyParameterTests.cs.
+public partial class RefReadonlyParameterTests : CSharpTestBase
+{
+    private const string RequiresLocationAttributeName = "RequiresLocationAttribute";
+    private const string RequiresLocationAttributeNamespace = "System.Runtime.CompilerServices";
+    private const string RequiresLocationAttributeQualifiedName = $"{RequiresLocationAttributeNamespace}.{RequiresLocationAttributeName}";
+
+    private static void VerifyRequiresLocationAttributeSynthesized(ModuleSymbol module)
+    {
+        var attributeType = module.GlobalNamespace.GetMember<NamedTypeSymbol>(RequiresLocationAttributeQualifiedName);
+        if (module is SourceModuleSymbol)
+        {
+            Assert.Null(attributeType);
+        }
+        else
+        {
+            Assert.NotNull(attributeType);
+        }
+    }
+
+    private static void VerifyRefReadonlyParameter(ParameterSymbol parameter,
+        bool refKind = true,
+        bool metadataIn = true,
+        bool attributes = true,
+        bool modreq = false,
+        bool useSiteError = false)
+    {
+        if (refKind)
+        {
+            Assert.Equal(RefKind.RefReadOnlyParameter, parameter.RefKind);
```

Comment:
When `refKind` is false, we do no check. In that case, could we tighten to check that the RefKind is not RefReadOnlyParameter?
`Assert.Equal(refKind, RefKind.RefReadOnlyParameter == parameter.RefKind);`